cdxev/validator/validate.py

```diff
@@ -22,16 +22,25 @@ def validate_sbom(
     file: Path,
     report_format: t.Optional[str],
     report_path: t.Optional[Path],
-    schema_type: str = "default",
-    filename_regex: t.Optional[str] = "",
-    schema_path: str = "",
+    schema_type: t.Optional[str],
+    filename_regex: t.Optional[str],
+    schema_path: t.Optional[Path],
 ) -> int:
     errors = []
-    if input_format == "json":
-        sbom_schema, used_schema_path = open_schema(
-            sbom, file, schema_type, schema_path
+    if (schema_path is not None) == bool(schema_type):
+        raise AssertionError(
+            "Exactly one of schema_path or schema_type must be non-None"
         )
 
+    if input_format == "json":
+        try:
+            spec_version = sbom["specVersion"]
+        except KeyError:
+            logger.info("SBOM doesn't specify CycloneDX version. Assuming version 1.3.")
```

Comment:
Do we really want to have this info? As `specVersion` is required by the "vanilla" specification, too, we would now get the same information twice:
![image](https://github.com/user-attachments/assets/ddc760ad-5cb0-48d7-b654-88a89f5d3c31)

Maybe this should be on `debug`-level?

Reply:
I hadn't even considered the fact that `specVersion` is mandatory.
Well, in any case, if we default to a version if it's not specified, then IMO we should inform the user. Otherwise, they might be confused by validation results that aren't applicable to the CDX version they had in mind. To my mind, the only alternative would be to fail early if the version is missing.

Reply:
"fail early" seems like the wrong decision as we would require an additional validation sequence. And as the "vanilla" specification requires it too, all the tools are also include the field. So the only scenario where a missing `specVersion` could happen, is in case of a manually created SBOM. Maybe we can catch the error from the validation and modify it, like we already do for e.g. [SPDX-IDs](https://github.com/Festo-se/cyclonedx-editor-validator/blob/0369c2143339be5ee03e146e63738c83e5e92577/cdxev/validator/validate.py#L127)?

Reply:
> "fail early" seems like the wrong decision as we would require an additional validation sequence. 

You mean the user would have to fix the error (i.e., add `specVersion`) and then rerun the validator? That's required anyways, unless they by chance targeted the same version that we fall back to. Otherwise, the results are unreliable.

> And as the "vanilla" specification requires it too, all the tools are also include the field. So the only scenario where a missing `specVersion` could happen, is in case of a manually created SBOM. Maybe we can catch the error from the validation and modify it, like we already do for e.g. [SPDX-IDs](https://github.com/Festo-se/cyclonedx-editor-validator/blob/0369c2143339be5ee03e146e63738c83e5e92577/cdxev/validator/validate.py#L127)?

You want to go to that effort for the 0.x% of use cases where somebody authors an SBOM manually and forgets the `specVersion` field? Be my guest, I'm out 😝

Reply:
> You mean the user would have to fix the error (i.e., add specVersion) and then rerun the validator? That's required anyways, unless they by chance targeted the same version that we fall back to. Otherwise, the results are unreliable.

Hmm, it does make sense. We even had once the use-case that colleagues manually created SBOMs and wrote `"specVersion": 1.4` but added fields from `1.5`. And then had to re-run the validation anyway. So: Fine for me with the "fail early", better than having the same message twice.

Reply:
Done